Add `optimize_for_size` variants for stable and unstable sort as well as select_nth_unstable

Voultapher opened the pull request:
- Stable sort uses a simple merge-sort that re-uses the existing - rather gnarly - merge function.
- Unstable sort jumps directly to the branchless heapsort fallback.
- select_nth_unstable jumps directly to the median_of_medians fallback, which is augmented with a custom tiny smallsort and partition impl.

Some code is duplicated but de-duplication would bring it's own problems. For example `swap_if_less` is critical for performance, if the sorting networks don't inline it perf drops drastically, however `#[inline(always)]` is also a poor fit, if the provided comparison function is huge, it gives the compiler an out to only instantiate `swap_if_less` once and call it. Another aspect that would suffer when making `swap_if_less` pub, is having to cfg out dozens of functions in in smallsort module.

Part of https://github.com/rust-lang/rust/issues/125612

r​? @Kobzol

Voultapher committed:
Add binary-size optimized variants for stable and unstable sort as well as select_nth_unstable

- Stable sort uses a simple merge-sort that re-uses the existing - rather gnarly - merge function.
- Unstable sort jumps directly to the branchless heapsort fallback.
- select_nth_unstable jumps directly to the median_of_medians fallback, which is augmented with a
  custom tiny smallsort and partition impl.

Some code is duplicated but de-duplication would bring it's own problems. For example `swap_if_less`
is critical for performance, if the sorting networks don't inline it perf drops drastically,
however `#[inline(always)]` is also a poor fit, if the provided comparison function is huge,
it gives the compiler an out to only instantiate `swap_if_less` once and call it. Another aspect
that would suffer when making `swap_if_less` pub, is having to cfg out dozens of functions in
in smallsort module.

Voultapher commented:
r? @Kobzol

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/10549332599/job/29224250958) [(plain)](https://github.com/rust-lang/rust/commit/13d7b546dada1cb2960dec90875b16a42ff7fb84/checks/29224250958/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
#16 2.865 Building wheels for collected packages: reuse
#16 2.866   Building wheel for reuse (pyproject.toml): started
#16 3.117   Building wheel for reuse (pyproject.toml): finished with status 'done'
#16 3.118   Created wheel for reuse: filename=reuse-4.0.3-cp310-cp310-manylinux_2_35_x86_64.whl size=132715 sha256=dfa09868353292d98f811d3efdb0d54d07389e808efc71d68e3b93c514bf8bec
#16 3.119   Stored in directory: /tmp/pip-ephem-wheel-cache-nq13luie/wheels/3d/8d/0a/e0fc6aba4494b28a967ab5eaf951c121d9c677958714e34532
#16 3.121 Installing collected packages: boolean-py, binaryornot, tomlkit, reuse, python-debian, markupsafe, license-expression, jinja2, chardet, attrs
#16 3.532 Successfully installed attrs-23.2.0 binaryornot-0.4.4 boolean-py-4.0 chardet-5.2.0 jinja2-3.1.4 license-expression-30.3.0 markupsafe-2.1.5 python-debian-0.1.49 reuse-4.0.3 tomlkit-0.13.0
#16 3.532 WARNING: Running pip as the 'root' user can result in broken permissions and conflicting behaviour with the system package manager. It is recommended to use a virtual environment instead: https://pip.pypa.io/warnings/venv
#16 DONE 3.6s
```

</details>

Kobzol commented:
r? libs

tgross35 commented on library/core/src/slice/sort/select.rs:
```diff
@@ -298,3 +328,92 @@ fn median_idx<T, F: FnMut(&T, &T) -> bool>(
+    }
+}
+
+#[cfg(feature = "optimize_for_size")]
+fn partition_size_opt<T, F>(v: &mut [T], pivot: usize, is_less: &mut F) -> usize
```
This likely actually applies a few other places. For example, the selection between `insertion_sort_shift_left` and `bubble_sort`, among others.

These actual implementation functions could be kept relatively private, then introduce new functions `sort_impl`/`sort_unstable_impl` that more public and just selects between the two with `cfg`. Tests should test `sort_impl` rather than the individual functions, and then anything that needs it can just call `sort_impl` without duplicating the config.

This way we more or less know that the implementations are the same (they pass the same tests), and anything that calls it doesn't need to worry about `cfg` or what the actual implementation is (not that I think these get called in multiple places, but they could).

Voultapher replied:
As said above, I think it's a bad idea to test these private implementation details, and so they should stay private. I prefer different names because it makes it clear that logic diverges here, Rust cares a lot about explicitness and I'd prefer to make it very clear to future readers that logic diverges depending on `optimize_for_size` where it does.

That said, thinking about it a bit more, I think I found a way to reduce code duplication in this specific place.

Voultapher commented:
One thing that is still outstanding is, how do we document the different implementation behavior?

I think adding even more text to the 10 places that document the current implementation of `sort`, `sort_unstable` and `select_nth_unstable` isn't a good idea. Maybe there could be something in a specific Rust book that mentions the implications of using `optimize_for_size`.

That said I think with how recent and WIP `optimize_for_size` still is, I'd say it's not a blocker for this PR.

Voultapher commented on library/core/src/slice/sort/select.rs:
```diff
@@ -40,7 +44,15 @@ where
         v.swap(min_idx, index);
     } else {
-        partition_at_index_loop(v, index, None, &mut is_less);
+        #[cfg(not(feature = "optimize_for_size"))]
+        {
+            partition_at_index_loop(v, index, None, &mut is_less);
+        }
+
+        #[cfg(feature = "optimize_for_size")]
+        {
+            median_of_medians(v, &mut is_less, index);
+        }
```
I'm familiar with that possibility, however for me it falls into a similar category as no-block if and loop statements in C++, which I find to be an anti-pattern and have caused some [legendary bugs](https://dwheeler.com/essays/apple-goto-fail.html). So I'd prefer to keep it as blocks.

tgross35 replied:
We do also have `cfg_if!` in core to make the nondivergence clear.

Voultapher replied:
I wasn't aware of `cfg_if!` that's a good fit here, will update.

Voultapher commented on library/core/src/slice/sort/select.rs:
```diff
@@ -298,3 +328,92 @@ fn median_idx<T, F: FnMut(&T, &T) -> bool>(
+// would clutter that module with cfg statements and make it generally harder to read and develop.
+// So to decouple things and simplify it, we use a an even smaller bubble sort.
+#[cfg(feature = "optimize_for_size")]
+fn bubble_sort<T, F: FnMut(&T, &T) -> bool>(v: &mut [T], is_less: &mut F) {
```
I'm subscriber to the school of exclusively testing public interfaces, this function gets tested transitively by the existing tests.

Voultapher commented on library/core/src/slice/sort/select.rs:
```diff
@@ -298,3 +328,92 @@ fn median_idx<T, F: FnMut(&T, &T) -> bool>(
+        intrinsics::abort();
+    }
+
+    // SAFETY: We checked that `pivot` is in-bounds.
+    unsafe {
+        // Place the pivot at the beginning of slice.
+        v.swap_unchecked(0, pivot);
+    }
```
If I move the `// Place the pivot at the beginning of slice.` comment to outside the block I find it's less clear what part of the comment is safety related and what isn't.

Voultapher committed:
Reduce code duplication by moving partition_lomuto_branchless_simple into quicksort module

Voultapher committed:
Convert cfg blocks to cfg_if

cuviper commented on library/core/src/slice/sort/select.rs:
```diff
@@ -298,3 +314,26 @@ fn median_idx<T, F: FnMut(&T, &T) -> bool>(
+
+// It's possible to re-use the insertion sort in the smallsort module, but with optimize_for_size it
+// would clutter that module with cfg statements and make it generally harder to read and develop.
+// So to decouple things and simplify it, we use a an even smaller bubble sort.
```
nit: 
```suggestion
// So to decouple things and simplify it, we use an even smaller bubble sort.
```

cuviper commented on library/core/src/slice/sort/select.rs:
```diff
@@ -298,3 +314,26 @@ fn median_idx<T, F: FnMut(&T, &T) -> bool>(
+            unsafe {
+                if is_less(v.get_unchecked(i), v.get_unchecked(i - 1)) {
+                    v.swap_unchecked(i - 1, i);
+                    did_swap = true;
```
Did you consider the "last swap" optimization instead? (mentioned [here](https://en.wikipedia.org/wiki/Bubble_sort#Optimizing_bubble_sort))

Voultapher replied:
Good idea, I think that shouldn't change binary-size much, let me try it.

Voultapher replied:
Together with `ptr::swap_nonoverlapping` we can cut it down from [38](https://godbolt.org/z/rzGcsMvds) instructions to [28](https://godbolt.org/z/4n7Eqe4YP). `ptr::swap_nonoverlapping` is also better for larger types, plus not having to call a function on each loop iteration is better.

cuviper commented on library/core/src/slice/sort/stable/tiny.rs:
```diff
@@ -0,0 +1,75 @@
+
+/// Swap two values in the slice pointed to by `v_base` at the position `a_pos` and `b_pos` if the
+/// value at position `b_pos` is less than the one at position `a_pos`.
+unsafe fn swap_if_less<T, F>(v_base: *mut T, a_pos: usize, b_pos: usize, is_less: &mut F)
```
Is this a copy of the other `swap_if_less`? I don't see any difference off-hand, so I wonder why you forked it.

Voultapher replied:
I'm not super happy either, but I think it's the best compromise as explained in the PR description:

> Some code is duplicated but de-duplication would bring it's own problems. For example swap_if_less is critical for performance, if the sorting networks don't inline it perf drops drastically, however #[inline(always)] is also a poor fit, if the provided comparison function is huge, it gives the compiler an out to only instantiate swap_if_less once and call it. Another aspect that would suffer when making swap_if_less pub, is having to cfg out dozens of functions in in smallsort module.

cuviper replied:
I agree that `#[inline(always)]` is troublesome, but wouldn't a regular `#[inline]` hint suffice? Although given the generic `F`, I would mostly expect this to get monomorphized locally anyway, making it a good inline candidate automatically.

I think the cfg issue could be solved by moving `swap_if_less` to a more common place, no?

Voultapher replied:
I don't feel comfortable with risking the optimizations breaking now in some un-tested scenario or the future or risking huge binary-sizes for the price of repeating 10 lines of code.

cuviper commented on library/core/src/slice/sort/unstable/quicksort.rs:
```diff
@@ -98,13 +101,15 @@ where
+    // SAFETY: We checked that `pivot` is in-bounds.
+    unsafe {
+        // Place the pivot at the beginning of slice.
+        v.swap_unchecked(0, pivot);
```
Is the abort above no longer sufficient to prove in-bounds for some reason?

Voultapher replied:
I'm afraid it never was, I looked at this code again with godbolt and the compiler didn't omit the panics before. IIRC I inspected that piece of code in a full binary disassembly in the past, not sure what my process error was, but doing `swap_unchecked` allows us to be sure, and guard against future compilers optimizing differently.

Voultapher committed:
Use last swap optimization in bubblesort

Voultapher committed:
Use simpler branchy swap logic in tiny merge sort

Avoids the code duplication issue and results in
smaller binary size, which after all is the
purpose of the feature.

Voultapher commented:
I think the open points have been addressed, if not, please re-open. Good to be merged from my side.

Voultapher commented:
Would be nice to measure the binary-size impact of this change in a real setting, here are the result of our tool that we also used in the design documents:

`slice::sort`:

Configuration                | Type     | Size default (bytes) | Size `optimize_for_size` (bytes)
-----------------------------|----------|----------------------|-------------------------------
release                      | `u64`    | 5304                 | 602
release                      | `String` | 6630                 | 838
release_lto_thin             | `u64`    | 5317                 | 570
release_lto_thin             | `String` | 6607                 | 806
release_lto_thin_opt_level_s | `u64`    | 3748                 | 584
release_lto_thin_opt_level_s | `String` | 5448                 | 813

`slice::sort_unstable`:

Configuration                | Type     | Size default (bytes) | Size `optimize_for_size` (bytes)
-----------------------------|----------|----------------------|-------------------------------
release                      | `u64`    | 3390                 | 264
release                      | `String` | 4972                 | 542
release_lto_thin             | `u64`    | 3390                 | 264
release_lto_thin             | `String` | 4972                 | 542
release_lto_thin_opt_level_s | `u64`    | 3070                 | 229
release_lto_thin_opt_level_s | `String` | 4339                 | 507

Voultapher committed:
Drop bubble_sort

While using it results in slightly slammer binaries, it's not deemed
worth it to add yet another sort algorithm to the standard library.
select_nth_unstable has bigger binary-size problems.

Voultapher committed:
Shrink heapsort further by combining sift_down loops

Voultapher committed:
Select tiny sorts for 16-bit platforms

Also skips stack alloc in stable sort if 16-bit target platform.

Voultapher committed:
Use non-overlapping swap for inner heapsort loop

This regresses binary-size slightly for normal builds, but the important
release_lto_thin_opt_level_s config sees a small improvement in
binary-size and a larger types such as string and 1k see 2-3% run-time
improvements with this change.

Voultapher commented:
Pushed a bunch of changes, and updated the table above.

cuviper commented:
LGTM, and thanks for the detailed tables and graphs as well!

@bors r+ rollup=never
(just to be sure of independent perf)

bors commented:
:pushpin: Commit 54391983486d5271c378821c8e6a8f018dfdd14d has been approved by `cuviper`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=cuviper 54391983486d5271c378821c8e6a8f018dfdd14d -->
<!-- homu: {"type":"Approved","sha":"54391983486d5271c378821c8e6a8f018dfdd14d","approver":"cuviper","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

bors commented:
:hourglass: Testing commit 54391983486d5271c378821c8e6a8f018dfdd14d with merge 363ae4188316b8b22cf6c1890bc73d84d05f70a4...
<!-- homu: {"type":"BuildStarted","head_sha":"54391983486d5271c378821c8e6a8f018dfdd14d","merge_sha":"363ae4188316b8b22cf6c1890bc73d84d05f70a4"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/11020051865/job/30610292131)
Approved by: cuviper
Pushing 363ae4188316b8b22cf6c1890bc73d84d05f70a4 to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"cuviper","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/11020051865/job/30610292131"},"merge_sha":"363ae4188316b8b22cf6c1890bc73d84d05f70a4"} -->

rust-timer commented:
Finished benchmarking commit (363ae4188316b8b22cf6c1890bc73d84d05f70a4): [comparison URL](https://perf.rust-lang.org/compare.html?start=67bb749c2e1cf503fee64842963dd3e72a417a3f&end=363ae4188316b8b22cf6c1890bc73d84d05f70a4&stat=instructions:u).

### Overall result: ❌ regressions - ACTION NEEDED


**Next Steps**: If you can justify the regressions found in this perf run, please indicate this with `@rustbot label: +perf-regression-triaged` along with sufficient written justification. If you cannot justify the regressions please open an issue or create a new PR that fixes the regressions, add a comment linking to the newly created issue or PR, and then add the `perf-regression-triaged` label to this PR.

@rustbot label: +perf-regression
cc @rust-lang/wg-compiler-performance



### [Instruction count](https://perf.rust-lang.org/compare.html?start=67bb749c2e1cf503fee64842963dd3e72a417a3f&end=363ae4188316b8b22cf6c1890bc73d84d05f70a4&stat=instructions:u)
This is a highly reliable metric that was used to determine the overall result at the top of this comment.

|                                    | mean | range        | count |
|:----------------------------------:|:----:|:------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 0.3% | [0.2%, 0.3%] | 3     |
| Regressions ❌ <br /> (secondary)  | -    | -            | 0     |
| Improvements ✅ <br /> (primary)   | -    | -            | 0     |
| Improvements ✅ <br /> (secondary) | -    | -            | 0     |
| All ❌✅ (primary)                 | 0.3% | [0.2%, 0.3%] | 3     |

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=67bb749c2e1cf503fee64842963dd3e72a417a3f&end=363ae4188316b8b22cf6c1890bc73d84d05f70a4&stat=max-rss)
<details>
<summary>Results (primary 5.7%, secondary 2.6%)</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean | range        | count |
|:----------------------------------:|:----:|:------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 5.7% | [4.0%, 9.1%] | 3     |
| Regressions ❌ <br /> (secondary)  | 2.6% | [2.0%, 3.1%] | 3     |
| Improvements ✅ <br /> (primary)   | -    | -            | 0     |
| Improvements ✅ <br /> (secondary) | -    | -            | 0     |
| All ❌✅ (primary)                 | 5.7% | [4.0%, 9.1%] | 3     |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=67bb749c2e1cf503fee64842963dd3e72a417a3f&end=363ae4188316b8b22cf6c1890bc73d84d05f70a4&stat=cycles:u)
<details>
<summary>Results (primary -0.5%, secondary 3.8%)</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | 3.8%  | [2.8%, 4.3%]   | 5     |
| Improvements ✅ <br /> (primary)   | -0.5% | [-0.5%, -0.5%] | 1     |
| Improvements ✅ <br /> (secondary) | -     | -              | 0     |
| All ❌✅ (primary)                 | -0.5% | [-0.5%, -0.5%] | 1     |
</details>

### [Binary size](https://perf.rust-lang.org/compare.html?start=67bb749c2e1cf503fee64842963dd3e72a417a3f&end=363ae4188316b8b22cf6c1890bc73d84d05f70a4&stat=size:linked_artifact)
<details>
<summary>Results (primary 0.0%, secondary -0.1%)</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 0.2%  | [0.0%, 0.8%]   | 14    |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -0.1% | [-0.1%, -0.0%] | 21    |
| Improvements ✅ <br /> (secondary) | -0.1% | [-0.2%, -0.1%] | 38    |
| All ❌✅ (primary)                 | 0.0%  | [-0.1%, 0.8%]  | 35    |
</details>

**Bootstrap**: 767.328s -> 768.432s (0.14%)
**Artifact size**: 340.88 MiB -> 340.84 MiB (-0.01%)

RalfJung commented on library/core/src/slice/sort/stable/mod.rs:
```diff
@@ -1,15 +1,24 @@
+#[cfg(not(feature = "optimize_for_size"))]
 pub(crate) mod quicksort;
 
+#[cfg(feature = "optimize_for_size")]
```
This likely needs to be `#[cfg(any(feature = "optimize_for_size", target_pointer_width = "16"))]`, see https://github.com/rust-lang/rust/issues/130818

Voultapher replied:
Good catch.

Voultapher commented:
Not sure the rust-timer run tells us anything useful here, given that all changes here happened to code behind a libcore feature that isn't being built IIUC.

Kobzol commented:
I don't think that further investigation is needed. This has seemingly perturbed codegen a little bit, but otherwise it's hidden behind an optional flag. The `image` regression has flipped back right after.

@rustbot label: +perf-regression-triaged
